File: Decorator/decorator.rb
class Burger 
    attr_accessor :description, :cost
    def initialize
        @description = "Burger"
        @cost = 5
    end 

end

class Burger_Decorator
    attr_accessor :description, :cost
    def initialize(burger,description,cost)
        @burger = burger
        @description = "#{@burger.description} with #{description}" 
        @cost = @burger.cost + cost
    end
end

#Burger
burger = Burger.new
burger.description
burger.cost

#Cheeseburger
cheeseburger = Burger_Decorator.new(burger,"cheese",0.50)
cheeseburger.description 
cheeseburger.cost

#CheeseBurger with Ketchup
ketchup_cheeseburger = Burger_Decorator.new(cheeseburger,"ketchup",0.10)
ketchup_cheeseburger.description
ketchup_cheeseburger.cost

File: Decorator/decorator.js
//Base Class
var Burger = function(){
    this.description = "Burger";  
};
Burger.prototype.getCost = function(){
    return 5.00;  
};

//Decorator Class
var BurgerDecorator = function(burger,description,cost){
    this.burger = burger; //instance of burger to be decorated
    this.description = this.burger.description + " with " + description;
    this.cost = cost;
};
BurgerDecorator.prototype.getCost = function(){
    return this.burger.getCost() + this.cost;
};


var burger = new Burger();
//Decorate Burger with Cheese
var cheeseBurger = new BurgerDecorator(burger,"Cheese", .20);
console.log(cheeseBurger.description);
console.log(cheeseBurger.getCost());

//Decorate Cheeseburger with ketchup
var cheeseBurgerWithKetchup = new BurgerDecorator(cheeseBurger,"Ketchup",.10);

console.log(cheeseBurgerWithKetchup.description);
console.log(cheeseBurgerWithKetchup.getCost());




